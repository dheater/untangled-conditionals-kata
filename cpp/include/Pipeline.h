
#include "Config.h"
#include "Emailer.h"
#include "Logger.h"
#include "Project.h"

class Pipeline {
    Config &config;
    Emailer &emailer;
    Logger &log;

    void send_email_if_enabled(const std::string& msg) const;
    void deploy(bool succeeded) const;

  public:
    Pipeline(Config &config, Emailer &emailer, Logger &log)
        : config(config)
        , emailer(emailer)
        , log(log)
    {
    }

    void run(Project project);
};
