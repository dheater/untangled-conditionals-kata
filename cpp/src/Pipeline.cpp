#include "Pipeline.h"

void Pipeline::run(Project project)
{
    if(!project.hasTests()) {
        log.info("No tests");
        deploy(project);
        return;
    }

    if(!("success" == project.runTests())) {
        log.error("Tests failed");
        send_email_if_enabled("Tests failed");
        return;
    }

    log.info("Tests passed");
    deploy(project);
}

void Pipeline::deploy(Project &project) const
{
    if(!("success" == project.deploy())) {
        log.error("Deployment failed");
        send_email_if_enabled("Deployment failed");
        return;
    }

    log.info("Deployment successful");
    send_email_if_enabled("Deployment completed successfully");
}

void Pipeline::send_email_if_enabled(const std::string &msg) const
{
    if(!config.sendEmailSummary()) {
        log.info("Email disabled");
        return;
    }

    log.info("Sending email");
    emailer.send(msg);
}
