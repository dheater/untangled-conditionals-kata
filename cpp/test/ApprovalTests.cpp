#include "CapturingEmailer.h"
#include "CapturingLogger.h"
#include <Config.h>
#include <Pipeline.h>

#include "ApprovalTests.hpp"
#include "catch2/catch.hpp"

static std::string run(bool email_enabled, bool build_successful,
                       TestStatus test_status)
{
    // Application setup
    Config config(email_enabled);
    CapturingEmailer emailer;
    CapturingLogger logger;

    Pipeline pipeline(config, emailer, logger);
    Project project(build_successful, test_status);

    // Run the scenario
    pipeline.run(project);

    // Get results
    std::string output = "\n";
    for(const std::string &msg : logger.getLoggedLines()) {
        output += msg + "\n";
    }
    for(const std::string &mail : emailer.getMail()) {
        output += mail + "\n";
    }

    return output;
}

TEST_CASE("PipelineApprovals")
{
    // Permutations
    std::vector<bool> email_enabled{ true, false };
    std::vector<bool> build_successful{ true, false };
    std::vector<TestStatus> test_status{ TestStatus::PASSING_TESTS,
                                         TestStatus::FAILING_TESTS,
                                         TestStatus::NO_TESTS };

    ApprovalTests::CombinationApprovals::verifyAllCombinations(
        [](bool email_enabled, bool build_successful, TestStatus test_status) {
            return run(email_enabled, build_successful, test_status);
        },
        email_enabled, build_successful, test_status);
}

inline std::ostream &operator<<(std::ostream &os, TestStatus ts)
{
    switch(ts) {
    case NO_TESTS:
        os << "No Tests";
        break;
    case PASSING_TESTS:
        os << "Tests Passed";
        break;
    case FAILING_TESTS:
        os << "Tests Failed";
        break;
    }
    return os;
}

