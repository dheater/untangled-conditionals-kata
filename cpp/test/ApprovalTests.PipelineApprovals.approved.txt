(1, 1, Tests Passed) => 
INFO: Tests passed
INFO: Deployment successful
INFO: Sending email
Deployment completed successfully

(1, 1, Tests Failed) => 
ERROR: Tests failed
INFO: Sending email
Tests failed

(1, 1, No Tests) => 
INFO: No tests
INFO: Deployment successful
INFO: Sending email
Deployment completed successfully

(1, 0, Tests Passed) => 
INFO: Tests passed
ERROR: Deployment failed
INFO: Sending email
Deployment failed

(1, 0, Tests Failed) => 
ERROR: Tests failed
INFO: Sending email
Tests failed

(1, 0, No Tests) => 
INFO: No tests
ERROR: Deployment failed
INFO: Sending email
Deployment failed

(0, 1, Tests Passed) => 
INFO: Tests passed
INFO: Deployment successful
INFO: Email disabled

(0, 1, Tests Failed) => 
ERROR: Tests failed
INFO: Email disabled

(0, 1, No Tests) => 
INFO: No tests
INFO: Deployment successful
INFO: Email disabled

(0, 0, Tests Passed) => 
INFO: Tests passed
ERROR: Deployment failed
INFO: Email disabled

(0, 0, Tests Failed) => 
ERROR: Tests failed
INFO: Email disabled

(0, 0, No Tests) => 
INFO: No tests
ERROR: Deployment failed
INFO: Email disabled


